File: bin/IT6/ChannelCapacity.m
function [s,Capacity]=ChannelCapacity(n,m,py_x)
%Input  -n 信源符号个数
%       -m 信宿符号个数
%       -py_x n*m矩阵，信道转移概率矩阵
%Output -s 1*n矩阵 最佳信源分布
s=ones(1,n)/n;
C0=0;
C1=0;

while (1)
    for j=1:m
       for i=1:n
          a(j,i)=s(i)*py_x(i,j);
       end
    end
    
    temp=sum(a,2);
    
    for j=1:m
       for i=1:n
          phi(j,i)=a(j,i)/temp(i);
       end
    end
    
    
    
    for i=1:n
       s(i) =exp(py_x(i,:)*log2(phi(:,i)));
    end
    temp=sum(s);
    s=s./temp;
    
    temp1=0;
    for i=1:n
        temp2=0;
        for j=1:m
            temp2=temp2+py_x(i,j)*log2(phi(j,i));
        end
        temp1=temp1+exp(temp2);
    end
    C0=C1;C1=log2(temp1);
    
    if (C1-C0<1e-6)
        break;
    end   
    
end
Capacity=C1;

File: src/IT6/ChannelCapacity.java
package IT6;

public class ChannelCapacity {
	protected int N,M;
	protected int time;
	protected double C;
	protected double[][] p;
	protected double[] S;
	protected double[] SS;
	
	
	public ChannelCapacity() {
		
	}
	
	
	public ChannelCapacity(int N,int M,double [][] p) {
		this.N=N;
		this.M=M;
		this.p=p;
		S=new double[N];
		SS=new double[N];
	}
	
	public void compute() {
		
		
		/** 初始化数据 */
		for (int i = 0; i < N; i++)
			S[i] = (double) 1 / N;// 赋值

		/** 迭代 */
		boolean flag = true;
		double[][] phi = new double[M][N];

		while (flag) {
			/** 首先计算phi ji */
			for (int j = 0; j < M; j++) {
				double sum = 0;
				for (int i = 0; i < N; i++) {
					sum = sum + S[i] * p[i][j];
				}
				for (int i = 0; i < N; i++)
					phi[j][i] = (S[i] * p[i][j]) / sum;
			}
			/** 迭代计算S */

			// 计算分母
			double sum1 = 0;
			for (int i = 0; i < N; i++) {
				boolean flag3 = true;
				double sum2 = 0;
				for (int j = 0; j < M; j++) {
					if (phi[j][i] != 0)
						sum2 = sum2
								+ (p[i][j] * (Math.log(phi[j][i]) / Math
										.log(Math.E)));
					else if (phi[j][i] == 0 && p[i][j] != 0) {
						flag3 = false;
					}// exp(log0)=0,下同
					else if (phi[j][i] == 0 && p[i][j] == 0)
						sum2 = sum2 + 0;// 0log0=0
				}

				if (flag3)
					sum1 = sum1 + Math.exp(sum2);
				else
					sum1 = sum1 + 0;
			}

			/** 计算SS[i] */

			for (int i = 0; i < N; i++) {
				boolean flag1 = true;// 若有无穷比无穷
				double sum6 = 0;
				for (int j = 0; j < M; j++) {
					if (phi[j][i] != 0)
						sum6 = sum6 + p[i][j]
								* (Math.log(phi[j][i]) / Math.log(Math.E));
					else if (phi[j][i] == 0 && p[i][j] != 0) {
						flag1 = false;
					} else if (phi[j][i] == 0 && p[i][j] == 0)
						sum6 = sum6 + 0;
				}
				if (flag1)
					SS[i] = Math.exp(sum6) / sum1;
				else
					SS[i] = 0;
			}
			double distance = 0;
			for (int i = 0; i < N; i++) {
				distance = distance + Math.pow(SS[i] - S[i], 2);// 计算范数
			}
			if (distance < 0.00001)
				flag = false;
			else {
				for (int i = 0; i < N; i++)
					S[i] = SS[i];
			}
			C = Math.log(sum1) / Math.log(2);
			time++;
		}
		
	}
	
	public void display() {
		System.out.println("迭代次数："+time);
		System.out.print("最佳信源分布：(");
		for (int i = 0; i < N - 1; i++)	System.out.printf("%5.3f, ",S[i]);
		System.out.printf("%5.3f",S[N - 1]);
		System.out.println(")");
		System.out.printf("信道容量：%8.3f\n" ,C);
		
	}
	
	
	
}
